codec-http/src/main/java/io/netty/handler/codec/http/router/MethodRouted.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright 2014 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.codec.http.router;
+
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.util.ReferenceCounted;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class MethodRouted<T> implements ReferenceCounted {
+    private final T                         target;
+    private final boolean                   notFound;
+    private final HttpRequest               request;
+    private final String                    path;
+    private final Map<String, String>       pathParams;
+    private final Map<String, List<String>> queryParams;
+
+    private final ReferenceCounted requestAsReferenceCounted;
+
+    //--------------------------------------------------------------------------
+
+    public MethodRouted(
+        T                         target,
+        boolean                   notFound,
+        HttpRequest               request,
+        String                    path,
+        Map<String, String>       pathParams,
+        Map<String, List<String>> queryParams
+    ) {
+        this.target      = target;
+        this.notFound    = notFound;
+        this.request     = request;
+        this.path        = path;
+        this.pathParams  = pathParams;
+        this.queryParams = queryParams;
+
+        requestAsReferenceCounted = (request instanceof ReferenceCounted)? (ReferenceCounted) request : null;
+    }
+
+    public T target() {
+        return target;
+    }
+
+    public boolean notFound() {
```

Comment:
The naming is "on purpose" to keep the vocabulary consistent. For example, to set 404 handler:

```java
router.notFound(My404Handler.class);
```